### PR TITLE
Add a text to `unreachable!()`

### LINKED ISSUE #4959: Pedantic lint against `unreachable!()`
It would be nice to have a lint, that detects any use of the [`unreachable!`](https://doc.rust-lang.org/1.3.0/std/macro.unreachable!.html) macro without arguments.

This is useful for procedural macros, where only the following is shown
```
error: proc-macro derive panicked
 --> tests/test_union.rs:3:10
  |
3 | #[derive(ShortHand)]
  |          ^^^^^^^^^
  |
  = help: message: internal error: entered unreachable code

error: aborting due to previous error

error: could not compile `shorthand`.

To learn more, run the command again with --verbose.
```
This is fine, if you only have one or two uses of `unreachable`, but it becomes very difficult in larger codebases, because you have to check every occurrence of `unreachable` in the codebase.

Adding some text to the macro will  be a lot more helpful!
`unreachable!("ident is neither enable nor disable")`
```
error: proc-macro derive panicked
 --> tests/test_union.rs:3:10
  |
3 | #[derive(ShortHand)]
  |          ^^^^^^^^^
  |
  = help: message: internal error: entered unreachable code: ident is neither enable nor disable

error: aborting due to previous error

error: could not compile `shorthand`.

To learn more, run the command again with --verbose.
```

I think `panic!` in a proc-macro has the same issue.

### PR DESCRIPTION
**Pull request by chansuke**
Fixes #4959.

### CODE REVIEW

File: tests/ui/regex.stderr [hunk 1]
```diff
@@ -1,171 +1,26 @@
...
-   |
-LL |     let raw_string_error = Regex::new(r#"[...//...]"#);
-   |                                              ^^
-
-error: trivial regex
-  --> $DIR/regex.rs:42:33
-   |
-LL |     let trivial_eq = Regex::new("^foobar$");
-   |                                 ^^^^^^^^^^
+note: candidate #1: /Users/YusukeAbe/code/rust/rust-clippy/target/debug/deps/libregex-a27846278b6a956a.rlib
```
Comment (JohnTitor) [inline tests/ui/regex.stderr]: Hm, it depends on your environment. Shouldn't be changed.
Maybe: `cargo clean; cargo test` resolves this failure.

File: tests/ui/option_map_unit_fn_unfixable.stderr [hunk 2]
```diff
@@ -1,3 +1,8 @@
+error[E0465]: multiple rlib candidates for `regex` found
+   |
+   = note: candidate #1: /Users/YusukeAbe/code/rust/rust-clippy/target/debug/deps/libregex-a27846278b6a956a.rlib
+   = note: candidate #2: /Users/YusukeAbe/code/rust/rust-clippy/target/debug/deps/libregex-18da0247c966b574.rlib
```
Comment (JohnTitor) [inline tests/ui/option_map_unit_fn_unfixable.stderr]: Same here.

File: tests/ui/for_loop_unfixable.stderr [hunk 3]
```diff
@@ -1,9 +1,14 @@
+error[E0465]: multiple rlib candidates for `regex` found
+   |
+   = note: candidate #1: /Users/YusukeAbe/code/rust/rust-clippy/target/debug/deps/libregex-a27846278b6a956a.rlib
+   = note: candidate #2: /Users/YusukeAbe/code/rust/rust-clippy/target/debug/deps/libregex-18da0247c966b574.rlib
```
Comment (JohnTitor) [inline tests/ui/for_loop_unfixable.stderr]: Same here.

File: tests/ui/result_map_unit_fn_unfixable.stderr [hunk 4]
```diff
@@ -1,3 +1,8 @@
+error[E0465]: multiple rlib candidates for `regex` found
+   |
+   = note: candidate #1: /Users/YusukeAbe/code/rust/rust-clippy/target/debug/deps/libregex-a27846278b6a956a.rlib
+   = note: candidate #2: /Users/YusukeAbe/code/rust/rust-clippy/target/debug/deps/libregex-18da0247c966b574.rlib
```
Comment (JohnTitor) [inline tests/ui/result_map_unit_fn_unfixable.stderr]: Same here.

File: tests/ui/regex.stderr (hunk 1 above)
Comment (phansch) [reply tests/ui/regex.stderr]: `cargo update` sometimes fixes this for me as well

File: clippy_lints/src/panic_unimplemented.rs [hunk 5]
```diff
@@ -110,8 +110,8 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for PanicUnimplemented {
                               "`todo` should not be present in production code");
                 } else if is_expn_of(expr.span, "unreachable").is_some() {
                     let span = get_outer_span(expr);
-                    span_lint(cx, UNREACHABLE, span,
-                              "`unreachable` should not be present in production code");
+                    span_help_and_lint(cx, UNREACHABLE, span,
+                              "`unreachable` should not be present in production code", "ident is neither enable nor disable");
```
Comment (flip1995) [inline clippy_lints/src/panic_unimplemented.rs]: That was not, what the issue was talking about. Anyway, the requested lint already exists, so I don't think there is anything that needs to be done here.

File: tests/ui/regex.stderr (hunk 1 above)
Comment (lzutao) [reply tests/ui/regex.stderr]: I usually do `rm target/debug/deps/libregex*`.

Comment (chansuke) [reply tests/ui/regex.stderr]: Thanks for the review. I will fix this.

File: clippy_lints/src/panic_unimplemented.rs (hunk 5 above)
Comment (chansuke) [reply clippy_lints/src/panic_unimplemented.rs]: Thank you for the review and sorry for the misunderstanding. I will fix this.

File: tests/ui/option_map_unit_fn_unfixable.stderr (hunk 2 above)
Comment (chansuke) [reply tests/ui/option_map_unit_fn_unfixable.stderr]: Thanks again.

**Comment by flip1995** [comment]
Ping from triage @chansuke Thanks for addressing this issue. Meanwhile this issue is closed though. If you want to fix the related issue #3528 feel free to open a new PR.

**Pull request closed without merging**